File: app/src/test/java/ai/elimu/kukariri/logic/SpacedRepetitionHelperTest.java
package ai.elimu.kukariri.logic;

import org.junit.Test;
import static org.junit.Assert.assertThat;
import static org.hamcrest.CoreMatchers.is;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import ai.elimu.model.v2.gson.analytics.WordAssessmentEventGson;
import ai.elimu.model.v2.gson.analytics.WordLearningEventGson;

public class SpacedRepetitionHelperTest {

    /**
     * Test a time _before_ the time of the initial review (4 minutes).
     */
    @Test
    public void testIsReviewPending_3MinutesAfter() {
        Calendar calendar3MinutesAgo = Calendar.getInstance();
        calendar3MinutesAgo.add(Calendar.MINUTE, -3);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar3MinutesAgo);

        List<WordAssessmentEventGson> wordAssessmentEventGsonList = new ArrayList<>();

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsonList), is(false));
    }

    /**
     * Test a time _after_ the time of the initial review (4 minutes).
     */
    @Test
    public void testIsReviewPending_5MinutesAfter() {
        Calendar calendar5MinutesAgo = Calendar.getInstance();
        calendar5MinutesAgo.add(Calendar.MINUTE, -5);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar5MinutesAgo);

        List<WordAssessmentEventGson> wordAssessmentEventGsonList = new ArrayList<>();

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsonList), is(true));
    }

    /**
     * Test a time _after_ the time of the initial review (4 minutes).
     */
    @Test
    public void testIsReviewPending_5MinutesAfter_1stReviewNotMastered() {
        Calendar calendar5MinutesAgo = Calendar.getInstance();
        calendar5MinutesAgo.add(Calendar.MINUTE, -5);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar5MinutesAgo);

        Calendar calendarNow = Calendar.getInstance();
        WordAssessmentEventGson wordAssessmentEventGson = new WordAssessmentEventGson();
        wordAssessmentEventGson.setWordId(1L);
        wordAssessmentEventGson.setTimestamp(calendarNow);
        wordAssessmentEventGson.setMasteryScore(0.00f);

        List<WordAssessmentEventGson> wordAssessmentEventGsonList = new ArrayList<>();
        wordAssessmentEventGsonList.add(wordAssessmentEventGson);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsonList), is(true));
    }

    /**
     * Test a time _after_ the time of the initial review (4 minutes).
     */
    @Test
    public void testIsReviewPending_5MinutesAfter_1stReviewMastered() {
        Calendar calendar5MinutesAgo = Calendar.getInstance();
        calendar5MinutesAgo.add(Calendar.MINUTE, -5);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar5MinutesAgo);

        Calendar calendarNow = Calendar.getInstance();
        WordAssessmentEventGson wordAssessmentEventGson = new WordAssessmentEventGson();
        wordAssessmentEventGson.setWordId(1L);
        wordAssessmentEventGson.setTimestamp(calendarNow);
        wordAssessmentEventGson.setMasteryScore(1.00f);

        List<WordAssessmentEventGson> wordAssessmentEventGsonList = new ArrayList<>();
        wordAssessmentEventGsonList.add(wordAssessmentEventGson);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsonList), is(false));
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    public void testIsReviewPending_17MinutesAfter_MasteryOnFirstTry() {
        Calendar calendar17MinutesAgo = Calendar.getInstance();
        calendar17MinutesAgo.add(Calendar.MINUTE, -17);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar17MinutesAgo);

        Calendar calendarNow = Calendar.getInstance();
        WordAssessmentEventGson wordAssessmentEventGson = new WordAssessmentEventGson();
        wordAssessmentEventGson.setWordId(1L);
        wordAssessmentEventGson.setTimestamp(calendarNow);
        wordAssessmentEventGson.setMasteryScore(1.00f);

        List<WordAssessmentEventGson> wordAssessmentEventGsons = new ArrayList<>();
        wordAssessmentEventGsons.add(wordAssessmentEventGson);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsons), is(false));
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    public void testIsReviewPending_17MinutesAfter_MasteryOnSecondTry() {
        Calendar calendar22MinutesAgo = Calendar.getInstance();
        calendar22MinutesAgo.add(Calendar.MINUTE, -22);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar22MinutesAgo);

        Calendar calendar18MinutesAgo = Calendar.getInstance();
        calendar18MinutesAgo.add(Calendar.MINUTE, -18);
        WordAssessmentEventGson wordAssessmentEventGsonFirst = new WordAssessmentEventGson();
        wordAssessmentEventGsonFirst.setWordId(1L);
        wordAssessmentEventGsonFirst.setTimestamp(calendar18MinutesAgo);
        wordAssessmentEventGsonFirst.setMasteryScore(0.00f);

        Calendar calendar17MinutesAgo = Calendar.getInstance();
        calendar17MinutesAgo.add(Calendar.MINUTE, -17);
        WordAssessmentEventGson wordAssessmentEventGsonSecond = new WordAssessmentEventGson();
        wordAssessmentEventGsonSecond.setWordId(1L);
        wordAssessmentEventGsonSecond.setTimestamp(calendar17MinutesAgo);
        wordAssessmentEventGsonSecond.setMasteryScore(1.00f);

        // Store assessment events in _descending_ order
        List<WordAssessmentEventGson> wordAssessmentEventGsons = new ArrayList<>();
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond);
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsons), is(true));
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    public void testIsReviewPending_17MinutesAfter_MasteryOnThirdTry() {
        Calendar calendar23MinutesAgo = Calendar.getInstance();
        calendar23MinutesAgo.add(Calendar.MINUTE, -23);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar23MinutesAgo);

        Calendar calendar19MinutesAgo = Calendar.getInstance();
        calendar19MinutesAgo.add(Calendar.MINUTE, -19);
        WordAssessmentEventGson wordAssessmentEventGsonFirst = new WordAssessmentEventGson();
        wordAssessmentEventGsonFirst.setWordId(1L);
        wordAssessmentEventGsonFirst.setTimestamp(calendar19MinutesAgo);
        wordAssessmentEventGsonFirst.setMasteryScore(0.00f);

        Calendar calendar18MinutesAgo = Calendar.getInstance();
        calendar18MinutesAgo.add(Calendar.MINUTE, -18);
        WordAssessmentEventGson wordAssessmentEventGsonSecond = new WordAssessmentEventGson();
        wordAssessmentEventGsonSecond.setWordId(1L);
        wordAssessmentEventGsonSecond.setTimestamp(calendar18MinutesAgo);
        wordAssessmentEventGsonSecond.setMasteryScore(0.00f);

        Calendar calendar17MinutesAgo = Calendar.getInstance();
        calendar17MinutesAgo.add(Calendar.MINUTE, -17);
        WordAssessmentEventGson wordAssessmentEventGsonThird = new WordAssessmentEventGson();
        wordAssessmentEventGsonThird.setWordId(1L);
        wordAssessmentEventGsonThird.setTimestamp(calendar17MinutesAgo);
        wordAssessmentEventGsonThird.setMasteryScore(1.00f);

        // Store assessment events in _descending_ order
        List<WordAssessmentEventGson> wordAssessmentEventGsons = new ArrayList<>();
        wordAssessmentEventGsons.add(wordAssessmentEventGsonThird);
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond);
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsons), is(true));
    }

    /**
     * Test a time _before_ the time of the 2nd review (16 minutes).
     */
    @Test
    public void testIsReviewPending_false_15MinutesAfterFirstMastery() {
        Calendar calendar60MinutesAgo = Calendar.getInstance();
        calendar60MinutesAgo.add(Calendar.MINUTE, -60);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar60MinutesAgo);

        Calendar calendar15MinutesAgo = Calendar.getInstance();
        calendar15MinutesAgo.add(Calendar.MINUTE, -15);
        WordAssessmentEventGson wordAssessmentEventGsonFirst = new WordAssessmentEventGson();
        wordAssessmentEventGsonFirst.setWordId(1L);
        wordAssessmentEventGsonFirst.setTimestamp(calendar15MinutesAgo);
        wordAssessmentEventGsonFirst.setMasteryScore(1.00f);

        List<WordAssessmentEventGson> wordAssessmentEventGsons = new ArrayList<>();
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsons), is(false));
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    public void testIsReviewPending_true_17MinutesAfterFirstMastery() {
        Calendar calendar60MinutesAgo = Calendar.getInstance();
        calendar60MinutesAgo.add(Calendar.MINUTE, -60);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar60MinutesAgo);

        Calendar calendar17MinutesAgo = Calendar.getInstance();
        calendar17MinutesAgo.add(Calendar.MINUTE, -17);
        WordAssessmentEventGson wordAssessmentEventGsonFirst = new WordAssessmentEventGson();
        wordAssessmentEventGsonFirst.setWordId(1L);
        wordAssessmentEventGsonFirst.setTimestamp(calendar17MinutesAgo);
        wordAssessmentEventGsonFirst.setMasteryScore(1.00f);

        List<WordAssessmentEventGson> wordAssessmentEventGsons = new ArrayList<>();
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsons), is(true));
    }

    /**
     * Test a time _before_ the time of the 3rd review (64 minutes).
     */
    @Test
    public void testIsReviewPending_false_63MinutesAfterSecondMastery() {
        Calendar calendar180MinutesAgo = Calendar.getInstance();
        calendar180MinutesAgo.add(Calendar.MINUTE, -180);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar180MinutesAgo);

        // 1st review (>=4 minutes after learning event)
        // 180 minutes ago - 120 minutes ago = 60 minutes passed
        Calendar calendar120MinutesAgo = Calendar.getInstance();
        calendar120MinutesAgo.add(Calendar.MINUTE, -120);
        WordAssessmentEventGson wordAssessmentEventGsonFirst = new WordAssessmentEventGson();
        wordAssessmentEventGsonFirst.setWordId(1L);
        wordAssessmentEventGsonFirst.setTimestamp(calendar120MinutesAgo);
        wordAssessmentEventGsonFirst.setMasteryScore(1.00f);

        // 2nd review (>=16 minutes after the 1st mastery)
        // 120 minutes ago - 63 minutes ago = 57
        Calendar calendar63MinutesAgo = Calendar.getInstance();
        calendar63MinutesAgo.add(Calendar.MINUTE, -63);
        WordAssessmentEventGson wordAssessmentEventGsonSecond = new WordAssessmentEventGson();
        wordAssessmentEventGsonSecond.setWordId(1L);
        wordAssessmentEventGsonSecond.setTimestamp(calendar63MinutesAgo);
        wordAssessmentEventGsonSecond.setMasteryScore(1.00f);

        // Store assessment events in _descending_ order
        List<WordAssessmentEventGson> wordAssessmentEventGsons = new ArrayList<>();
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond);
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst);

        // 3rd review (>=64 minutes after the 2nd mastery)
        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsons), is(false));
    }

    /**
     * Test a time _after_ the time of the 3rd review (64 minutes).
     */
    @Test
    public void testIsReviewPending_true_65MinutesAfterSecondMastery() {
        Calendar calendar180MinutesAgo = Calendar.getInstance();
        calendar180MinutesAgo.add(Calendar.MINUTE, -180);
        WordLearningEventGson wordLearningEventGson = new WordLearningEventGson();
        wordLearningEventGson.setWordId(1L);
        wordLearningEventGson.setTimestamp(calendar180MinutesAgo);

        Calendar calendar120MinutesAgo = Calendar.getInstance();
        calendar120MinutesAgo.add(Calendar.MINUTE, -120);
        WordAssessmentEventGson wordAssessmentEventGsonFirst = new WordAssessmentEventGson();
        wordAssessmentEventGsonFirst.setWordId(1L);
        wordAssessmentEventGsonFirst.setTimestamp(calendar120MinutesAgo);
        wordAssessmentEventGsonFirst.setMasteryScore(1.00f);

        Calendar calendar65MinutesAgo = Calendar.getInstance();
        calendar65MinutesAgo.add(Calendar.MINUTE, -65);
        WordAssessmentEventGson wordAssessmentEventGsonSecond = new WordAssessmentEventGson();
        wordAssessmentEventGsonSecond.setWordId(1L);
        wordAssessmentEventGsonSecond.setTimestamp(calendar65MinutesAgo);
        wordAssessmentEventGsonSecond.setMasteryScore(1.00f);

        // Store assessment events in _descending_ order
        List<WordAssessmentEventGson> wordAssessmentEventGsons = new ArrayList<>();
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond);
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst);

        assertThat(SpacedRepetitionHelper.isReviewPending(wordLearningEventGson, wordAssessmentEventGsons), is(true));
    }
}

File: app/src/test/java/ai/elimu/kukariri/logic/SpacedRepetitionHelperTest.kt
package ai.elimu.kukariri.logic

import ai.elimu.model.v2.gson.analytics.WordAssessmentEventGson
import ai.elimu.model.v2.gson.analytics.WordLearningEventGson
import org.hamcrest.CoreMatchers
import org.junit.Assert
import org.junit.Test
import java.util.Calendar

class SpacedRepetitionHelperTest {
    /**
     * Test a time _before_ the time of the initial review (4 minutes).
     */
    @Test
    fun testIsReviewPending_3MinutesAfter() {
        val calendar3MinutesAgo = Calendar.getInstance()
        calendar3MinutesAgo.add(Calendar.MINUTE, -3)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar3MinutesAgo

        val wordAssessmentEventGsonList: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsonList
            ), CoreMatchers.`is`<Boolean?>(false)
        )
    }

    /**
     * Test a time _after_ the time of the initial review (4 minutes).
     */
    @Test
    fun testIsReviewPending_5MinutesAfter() {
        val calendar5MinutesAgo = Calendar.getInstance()
        calendar5MinutesAgo.add(Calendar.MINUTE, -5)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar5MinutesAgo

        val wordAssessmentEventGsonList: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsonList
            ), CoreMatchers.`is`<Boolean?>(true)
        )
    }

    /**
     * Test a time _after_ the time of the initial review (4 minutes).
     */
    @Test
    fun testIsReviewPending_5MinutesAfter_1stReviewNotMastered() {
        val calendar5MinutesAgo = Calendar.getInstance()
        calendar5MinutesAgo.add(Calendar.MINUTE, -5)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar5MinutesAgo

        val calendarNow = Calendar.getInstance()
        val wordAssessmentEventGson = WordAssessmentEventGson()
        wordAssessmentEventGson.wordId = 1L
        wordAssessmentEventGson.timestamp = calendarNow
        wordAssessmentEventGson.masteryScore = 0.00f

        val wordAssessmentEventGsonList: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsonList.add(wordAssessmentEventGson)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsonList
            ), CoreMatchers.`is`<Boolean?>(true)
        )
    }

    /**
     * Test a time _after_ the time of the initial review (4 minutes).
     */
    @Test
    fun testIsReviewPending_5MinutesAfter_1stReviewMastered() {
        val calendar5MinutesAgo = Calendar.getInstance()
        calendar5MinutesAgo.add(Calendar.MINUTE, -5)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar5MinutesAgo

        val calendarNow = Calendar.getInstance()
        val wordAssessmentEventGson = WordAssessmentEventGson()
        wordAssessmentEventGson.wordId = 1L
        wordAssessmentEventGson.timestamp = calendarNow
        wordAssessmentEventGson.masteryScore = 1.00f

        val wordAssessmentEventGsonList: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsonList.add(wordAssessmentEventGson)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsonList
            ), CoreMatchers.`is`<Boolean?>(false)
        )
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    fun testIsReviewPending_17MinutesAfter_MasteryOnFirstTry() {
        val calendar17MinutesAgo = Calendar.getInstance()
        calendar17MinutesAgo.add(Calendar.MINUTE, -17)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar17MinutesAgo

        val calendarNow = Calendar.getInstance()
        val wordAssessmentEventGson = WordAssessmentEventGson()
        wordAssessmentEventGson.wordId = 1L
        wordAssessmentEventGson.timestamp = calendarNow
        wordAssessmentEventGson.masteryScore = 1.00f

        val wordAssessmentEventGsons: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsons.add(wordAssessmentEventGson)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsons
            ), CoreMatchers.`is`<Boolean?>(false)
        )
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    fun testIsReviewPending_17MinutesAfter_MasteryOnSecondTry() {
        val calendar22MinutesAgo = Calendar.getInstance()
        calendar22MinutesAgo.add(Calendar.MINUTE, -22)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar22MinutesAgo

        val calendar18MinutesAgo = Calendar.getInstance()
        calendar18MinutesAgo.add(Calendar.MINUTE, -18)
        val wordAssessmentEventGsonFirst = WordAssessmentEventGson()
        wordAssessmentEventGsonFirst.wordId = 1L
        wordAssessmentEventGsonFirst.timestamp = calendar18MinutesAgo
        wordAssessmentEventGsonFirst.masteryScore = 0.00f

        val calendar17MinutesAgo = Calendar.getInstance()
        calendar17MinutesAgo.add(Calendar.MINUTE, -17)
        val wordAssessmentEventGsonSecond = WordAssessmentEventGson()
        wordAssessmentEventGsonSecond.wordId = 1L
        wordAssessmentEventGsonSecond.timestamp = calendar17MinutesAgo
        wordAssessmentEventGsonSecond.masteryScore = 1.00f

        // Store assessment events in _descending_ order
        val wordAssessmentEventGsons: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond)
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsons
            ), CoreMatchers.`is`<Boolean?>(true)
        )
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    fun testIsReviewPending_17MinutesAfter_MasteryOnThirdTry() {
        val calendar23MinutesAgo = Calendar.getInstance()
        calendar23MinutesAgo.add(Calendar.MINUTE, -23)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar23MinutesAgo

        val calendar19MinutesAgo = Calendar.getInstance()
        calendar19MinutesAgo.add(Calendar.MINUTE, -19)
        val wordAssessmentEventGsonFirst = WordAssessmentEventGson()
        wordAssessmentEventGsonFirst.wordId = 1L
        wordAssessmentEventGsonFirst.timestamp = calendar19MinutesAgo
        wordAssessmentEventGsonFirst.masteryScore = 0.00f

        val calendar18MinutesAgo = Calendar.getInstance()
        calendar18MinutesAgo.add(Calendar.MINUTE, -18)
        val wordAssessmentEventGsonSecond = WordAssessmentEventGson()
        wordAssessmentEventGsonSecond.wordId = 1L
        wordAssessmentEventGsonSecond.timestamp = calendar18MinutesAgo
        wordAssessmentEventGsonSecond.masteryScore = 0.00f

        val calendar17MinutesAgo = Calendar.getInstance()
        calendar17MinutesAgo.add(Calendar.MINUTE, -17)
        val wordAssessmentEventGsonThird = WordAssessmentEventGson()
        wordAssessmentEventGsonThird.wordId = 1L
        wordAssessmentEventGsonThird.timestamp = calendar17MinutesAgo
        wordAssessmentEventGsonThird.masteryScore = 1.00f

        // Store assessment events in _descending_ order
        val wordAssessmentEventGsons: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsons.add(wordAssessmentEventGsonThird)
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond)
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsons
            ), CoreMatchers.`is`<Boolean?>(true)
        )
    }

    /**
     * Test a time _before_ the time of the 2nd review (16 minutes).
     */
    @Test
    fun testIsReviewPending_false_15MinutesAfterFirstMastery() {
        val calendar60MinutesAgo = Calendar.getInstance()
        calendar60MinutesAgo.add(Calendar.MINUTE, -60)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar60MinutesAgo

        val calendar15MinutesAgo = Calendar.getInstance()
        calendar15MinutesAgo.add(Calendar.MINUTE, -15)
        val wordAssessmentEventGsonFirst = WordAssessmentEventGson()
        wordAssessmentEventGsonFirst.wordId = 1L
        wordAssessmentEventGsonFirst.timestamp = calendar15MinutesAgo
        wordAssessmentEventGsonFirst.masteryScore = 1.00f

        val wordAssessmentEventGsons: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsons
            ), CoreMatchers.`is`<Boolean?>(false)
        )
    }

    /**
     * Test a time _after_ the time of the 2nd review (16 minutes).
     */
    @Test
    fun testIsReviewPending_true_17MinutesAfterFirstMastery() {
        val calendar60MinutesAgo = Calendar.getInstance()
        calendar60MinutesAgo.add(Calendar.MINUTE, -60)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar60MinutesAgo

        val calendar17MinutesAgo = Calendar.getInstance()
        calendar17MinutesAgo.add(Calendar.MINUTE, -17)
        val wordAssessmentEventGsonFirst = WordAssessmentEventGson()
        wordAssessmentEventGsonFirst.wordId = 1L
        wordAssessmentEventGsonFirst.timestamp = calendar17MinutesAgo
        wordAssessmentEventGsonFirst.masteryScore = 1.00f

        val wordAssessmentEventGsons: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsons
            ), CoreMatchers.`is`<Boolean?>(true)
        )
    }

    /**
     * Test a time _before_ the time of the 3rd review (64 minutes).
     */
    @Test
    fun testIsReviewPending_false_63MinutesAfterSecondMastery() {
        val calendar180MinutesAgo = Calendar.getInstance()
        calendar180MinutesAgo.add(Calendar.MINUTE, -180)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar180MinutesAgo

        // 1st review (>=4 minutes after learning event)
        // 180 minutes ago - 120 minutes ago = 60 minutes passed
        val calendar120MinutesAgo = Calendar.getInstance()
        calendar120MinutesAgo.add(Calendar.MINUTE, -120)
        val wordAssessmentEventGsonFirst = WordAssessmentEventGson()
        wordAssessmentEventGsonFirst.wordId = 1L
        wordAssessmentEventGsonFirst.timestamp = calendar120MinutesAgo
        wordAssessmentEventGsonFirst.masteryScore = 1.00f

        // 2nd review (>=16 minutes after the 1st mastery)
        // 120 minutes ago - 63 minutes ago = 57
        val calendar63MinutesAgo = Calendar.getInstance()
        calendar63MinutesAgo.add(Calendar.MINUTE, -63)
        val wordAssessmentEventGsonSecond = WordAssessmentEventGson()
        wordAssessmentEventGsonSecond.wordId = 1L
        wordAssessmentEventGsonSecond.timestamp = calendar63MinutesAgo
        wordAssessmentEventGsonSecond.masteryScore = 1.00f

        // Store assessment events in _descending_ order
        val wordAssessmentEventGsons: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond)
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst)

        // 3rd review (>=64 minutes after the 2nd mastery)
        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsons
            ), CoreMatchers.`is`<Boolean?>(false)
        )
    }

    /**
     * Test a time _after_ the time of the 3rd review (64 minutes).
     */
    @Test
    fun testIsReviewPending_true_65MinutesAfterSecondMastery() {
        val calendar180MinutesAgo = Calendar.getInstance()
        calendar180MinutesAgo.add(Calendar.MINUTE, -180)
        val wordLearningEventGson = WordLearningEventGson()
        wordLearningEventGson.wordId = 1L
        wordLearningEventGson.timestamp = calendar180MinutesAgo

        val calendar120MinutesAgo = Calendar.getInstance()
        calendar120MinutesAgo.add(Calendar.MINUTE, -120)
        val wordAssessmentEventGsonFirst = WordAssessmentEventGson()
        wordAssessmentEventGsonFirst.wordId = 1L
        wordAssessmentEventGsonFirst.timestamp = calendar120MinutesAgo
        wordAssessmentEventGsonFirst.masteryScore = 1.00f

        val calendar65MinutesAgo = Calendar.getInstance()
        calendar65MinutesAgo.add(Calendar.MINUTE, -65)
        val wordAssessmentEventGsonSecond = WordAssessmentEventGson()
        wordAssessmentEventGsonSecond.wordId = 1L
        wordAssessmentEventGsonSecond.timestamp = calendar65MinutesAgo
        wordAssessmentEventGsonSecond.masteryScore = 1.00f

        // Store assessment events in _descending_ order
        val wordAssessmentEventGsons: MutableList<WordAssessmentEventGson> =
            ArrayList<WordAssessmentEventGson>()
        wordAssessmentEventGsons.add(wordAssessmentEventGsonSecond)
        wordAssessmentEventGsons.add(wordAssessmentEventGsonFirst)

        Assert.assertThat<Boolean?>(
            SpacedRepetitionHelper.isReviewPending(
                wordLearningEventGson,
                wordAssessmentEventGsons
            ), CoreMatchers.`is`<Boolean?>(true)
        )
    }
}
